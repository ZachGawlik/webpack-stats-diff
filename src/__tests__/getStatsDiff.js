const { getStatsDiff } = require('../index');

test('handles blank inputs', () => {
  expect(getStatsDiff([], [])).toMatchSnapshot();
});

test('reports stat diffs between old and new assets', () => {
  const oldAssets = [
    { name: 'commons.js', size: 32768 },
    { name: 'logo.svg', size: 588 },
    { name: 'me.jpg', size: 1200000 },
    { name: 'main-site.js', size: 68000 },
    { name: 'other-page.js', size: 40000 }
  ];
  const newAssets = [
    { name: 'commons.js', size: 65536 },
    { name: 'Roboto-Regular.ttf', size: 176999 },
    { name: 'me.jpg', size: 1200000 },
    { name: 'main-site.js', size: 38000 },
    { name: 'other-page.js', size: 12345 }
  ];
  expect(getStatsDiff(oldAssets, newAssets)).toMatchSnapshot();
});

test('Matches bundles by name (string before first period) and ext', () => {
  const oldAssets = [
    { name: 'main.a83fjads83ja.js', size: 32768 },
    { name: 'main.23r8f2398sdz.css', size: 4000 },
    { name: 'form.i8aejf3iad.chunk.js', size: 5000 },
    { name: 'form.aefc83zl9a.css', size: 2000 }
  ];
  const newAssets = [
    { name: 'main.aweifa09eh3.js', size: 16384 },
    { name: 'main.jo821kjdsi9.css', size: 5000 },
    { name: 'form.awe9f3nads893oz.chunk.js', size: 6500 },
    { name: 'form.89aw3jiuawefh8d.css', size: 1500 }
  ];
  expect(getStatsDiff(oldAssets, newAssets)).toMatchObject({
    added: [],
    removed: [],
    bigger: [
      { name: 'form.js', oldSize: 5000, newSize: 6500 },
      { name: 'main.css', oldSize: 4000, newSize: 5000 }
    ],
    smaller: [
      {
        name: 'main.js',
        oldSize: 32768,
        newSize: 16384
      },
      { name: 'form.css', oldSize: 2000, newSize: 1500 }
    ],
    sameSize: []
  });
});

test('filters assets by ext config', () => {
  const oldAssets = [
    { name: 'commons.js', size: 32768 },
    { name: 'main-site.js', size: 68000 },
    { name: 'old-page.js', size: 8000 },
    { name: 'styles.css', size: 10000 },
    { name: 'logo.svg', size: 588 },
    { name: 'me.jpg', size: 1200000 }
  ];
  const newAssets = [
    { name: 'commons.js', size: 65536 },
    { name: 'main-site.js', size: 38000 },
    { name: 'styles.css', size: 10000 },
    { name: 'me.jpg', size: 1200000 },
    { name: 'Roboto-Regular.ttf', size: 176999 }
  ];

  expect(
    getStatsDiff(oldAssets, newAssets, { extensions: ['.js', 'css'] })
  ).toMatchObject({
    added: [],
    removed: [{ name: 'old-page.js' }],
    bigger: [{ name: 'commons.js' }],
    smaller: [{ name: 'main-site.js' }],
    sameSize: [{ name: 'styles.css' }],
    total: { name: 'Total' },
    extensions: ['.js', '.css']
  });
});

test('Sorts by greatest size differences first', () => {
  const newAssets = [
    { name: 'commons.js', size: 65536 },
    { name: 'new-page.js', size: 8000 },
    { name: 'main-site.js', size: 38000 }
  ];
  expect(getStatsDiff([], newAssets).added).toMatchObject([
    { name: 'commons.js' },
    { name: 'main-site.js' },
    { name: 'new-page.js' }
  ]);
});

describe('Marks an asset as sameSize if % change is below threshold', () => {
  const oldAssets = [
    { name: 'big-file-small-change.js', size: 100000 },
    { name: 'big-file-biggest-change.js', size: 100000 },
    { name: 'big-file-no-change.js', size: 100000 },
    { name: 'small-file-small-change.js', size: 1000 },
    { name: 'small-file-big-change.js', size: 1000 }
  ];
  const newAssets = [
    { name: 'big-file-small-change.js', size: 104000 },
    { name: 'big-file-biggest-change.js', size: 110001 },
    { name: 'big-file-no-change.js', size: 100000 },
    { name: 'small-file-small-change.js', size: 960 },
    { name: 'small-file-big-change.js', size: 949 }
  ];

  test('Default threshold is 5%', () => {
    expect(getStatsDiff(oldAssets, newAssets)).toMatchObject({
      bigger: [{ name: 'big-file-biggest-change.js' }],
      smaller: [{ name: 'small-file-big-change.js' }],
      sameSize: [
        { name: 'big-file-small-change.js' },
        { name: 'big-file-no-change.js' },
        { name: 'small-file-small-change.js' }
      ],
      threshold: 5
    });
  });

  test('Threshold can be adjusted by configuration', () => {
    expect(getStatsDiff(oldAssets, newAssets, { threshold: 0 })).toMatchObject({
      bigger: [
        { name: 'big-file-biggest-change.js' },
        { name: 'big-file-small-change.js' }
      ],
      smaller: [
        { name: 'small-file-big-change.js' },
        { name: 'small-file-small-change.js' }
      ],
      sameSize: [{ name: 'big-file-no-change.js' }],
      threshold: 0
    });

    expect(getStatsDiff(oldAssets, newAssets, { threshold: 10 })).toMatchObject(
      {
        bigger: [{ name: 'big-file-biggest-change.js' }],
        smaller: [],
        sameSize: [
          { name: 'big-file-small-change.js' },
          { name: 'big-file-no-change.js' },
          { name: 'small-file-small-change.js' },
          { name: 'small-file-big-change.js' }
        ],
        threshold: 10
      }
    );
  });
});
