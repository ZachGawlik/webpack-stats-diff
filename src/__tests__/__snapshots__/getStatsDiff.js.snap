// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles blank inputs 1`] = `
Object {
  "added": Array [],
  "bigger": Array [],
  "extensions": null,
  "removed": Array [],
  "sameSize": Array [],
  "smaller": Array [],
  "threshold": 5,
  "total": Object {
    "diff": 0,
    "diffPercentage": 0,
    "name": "Total",
    "newSize": 0,
    "oldSize": 0,
  },
}
`;

exports[`reports stat diffs between old and new assets 1`] = `
Object {
  "added": Array [
    Object {
      "diff": 176999,
      "diffPercentage": Infinity,
      "name": "Roboto-Regular.ttf",
      "newSize": 176999,
      "oldSize": 0,
    },
  ],
  "bigger": Array [
    Object {
      "diff": 32768,
      "diffPercentage": 100,
      "name": "commons.js",
      "newSize": 65536,
      "oldSize": 32768,
    },
  ],
  "extensions": null,
  "removed": Array [
    Object {
      "diff": -588,
      "diffPercentage": -100,
      "name": "logo.svg",
      "newSize": 0,
      "oldSize": 588,
    },
  ],
  "sameSize": Array [
    Object {
      "diff": 0,
      "diffPercentage": 0,
      "name": "me.jpg",
      "newSize": 1200000,
      "oldSize": 1200000,
    },
  ],
  "smaller": Array [
    Object {
      "diff": -30000,
      "diffPercentage": -44.11765,
      "name": "main-site.js",
      "newSize": 38000,
      "oldSize": 68000,
    },
    Object {
      "diff": -27655,
      "diffPercentage": -69.1375,
      "name": "other-page.js",
      "newSize": 12345,
      "oldSize": 40000,
    },
  ],
  "threshold": 5,
  "total": Object {
    "diff": 151524,
    "diffPercentage": 11.29633,
    "name": "Total",
    "newSize": 1492880,
    "oldSize": 1341356,
  },
}
`;
